### PR TITLE
Assignement PFE 2023

### PR DESCRIPTION
**Pull request by cyberkarim**
I'm Karim Majdi, candidate for an end-of-study internship at EVEAD. Here is my contribution for the given assignement. Can you review it please.

Thank you

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -1 +1,87 @@
...
+import pytesseract
+import time
+
+
+### input images reading
+def read_img_from_source():
+  images_array = []
+  absolute_input_path = os.path.abspath("../input")
+  input_images_nouns = os.listdir(absolute_input_path)
+  input_images_nouns.remove(".ipynb_checkpoints")
```
Comment (omarmhaimdat) [inline src/main.py:16]: Crashes if the file is not in the directory. To make sure you only read images you could have forced a check before:
```python
input_images_nouns = [input_images_noun for input_images_noun in input_images_nouns if input_images_noun.endswith((".png", ".jpg", ".jpeg"))]
```

File: src/main.py [hunk 2]
```diff
@@ -1 +1,87 @@
...
+
+### post-process output text produced by pytesseract
+def text_post_processing(images_predicted_text_content,input_files_names):
+    # python dictionary that stores image text content to be written in output
+    dictionary = {}
+    for i in range(0,len(images_predicted_text_content)):
+    ## text post-processing
+       splitted_content = images_predicted_text_content[i].split("\n")
+       # eliminate all space strings, empty strings, End_of_file sign
+       content = list(filter(("").__ne__, splitted_content))
```
Comment (omarmhaimdat) [inline src/main.py:51]: Too complex filtering could have been much simpler:

```python
conditions = ("", " ", "  ", "\f")
content = list(filter(lambda x: x not in conditions, splitted_content))
```

File: src/main.py [hunk 3]
```diff
@@ -1 +1,87 @@
...
+
+### pytesseract image text conversion
+def extract_text_content(processed_arrays):
+  images_predicted_text_content = []
+  for processed_img in processed_arrays:
+    images_predicted_text_content += [pytesseract.image_to_string(processed_img)]
+  return images_predicted_text_content
+
+### post-process output text produced by pytesseract
+def text_post_processing(images_predicted_text_content,input_files_names):
```
Comment (omarmhaimdat) [inline src/main.py:44]: A bit light in terms of postprocessing, since the results will be highly noisy with these types of images

**Review by omarmhaimdat — COMMENTED** [review]
This submission is a good attempt to solve the problem. However, I do have some comments:

- The code is not well formatted. You should use a linter to format your code. I recommend using black.
- The extraction of the text from the images is not very good. Using the same pipeline you could have obtained better results.
  - You could have tuned the parameters of tesseract to get better results.
  - Add a postprocessing language model to correct the detected text.
  - Detect some interest areas in the image and apply OCR only on these areas.
  - Many ideas are possible.
- Resizing the image to 500x500 is not a good idea. Many of the images are not square, so you will lose information.
- Normalizing the image without taking into account the color is not a good idea. You should use a color space that is more adapted to the problem or at least use the color distribution of the image.

In conclusion, I like your minimalistic approach to the problem. I wish you have made more efforts to improve the results. I also like that you have followed all the instructions and did a somewhat proper attempt of validating your results

**Comment by cyberkarim** [comment]
Hello Omar,

Yes, thank you for your prompt feedback. Indeed very insightful ideas.

-  I did try using the PSM parameter of the image_to_text function but
couldn't get a better result as I got an almost empty string afterwards. I
should have tried Text segmentation/detection, then applied OCR on these
specific areas since the useful information will be contained in these
areas without much noise as you suggested.
- Resizing all input images to 500*500 could have been wrong for
rectangular images. I should have instead augmented height and width
separately to keep rectangular dimensions (for example augment an image
from 200*300 (300*400)
- as for the normalization, I should have instead used standardisation (
that subtract the pixels mean and divide by standard deviation) to
normalize according to the distribution of pixels values. Standardisation
was a better idea since pixels values will be transformed according to the
overall distribution of all images. Standardisation is more useful for
domain-specific images.
- The post processing idea using a language model is an idea that I haven't
thought of, but could be very useful to correct the output text. at first I though you talked about a model that automatically corrects sentence, but in reality it's sequence model that will predict the next likely word or character (in some sort, correcting the word) which I already have knowledge about. I haven't used it  in the context of images before, therefore it's an interesting idea to correct the text provided by the ocr model.

In brief, It's a very interesting challenge given the varying nature of
publicity images that I would like to take if given the opportunity.

Kind regards

Le mar. 31 janv. 2023 à 10:39, Omar MHAIMDAT ***@***.***> a
écrit :

**Comment by omarmhaimdat** [comment]
Your answer will be taken into consideration, you will have my answer on Monday.

**Comment by cyberkarim** [comment]
concerning The post processing idea using a language model, at first I though you talked about a model that automatically corrects sentence, but in reality it's sequence model that will predict the next likely word or character (in some sort, correcting the word) which I already have knowledge about. I haven't used it  in the context of images before, therefore it's an interesting idea to correct the text provided by the ocr model.
